frontend/apps/reader/modules/readerview.lua

```diff
@@ -594,26 +609,61 @@ function ReaderView:drawXPointerSavedHighlight(bb, x, y)
             end
         end
     end
+    return colorful
 end
 
-function ReaderView:drawHighlightRect(bb, _x, _y, rect, drawer, draw_note_mark)
+function ReaderView:drawHighlightRect(bb, _x, _y, rect, drawer, color, draw_note_mark)
     local x, y, w, h = rect.x, rect.y, rect.w, rect.h
     if drawer == "lighten" then
-        bb:lightenRect(x, y, w, h, self.highlight.lighten_factor)
+        if not color then
+            bb:darkenRect(x, y, w, h, self.highlight.lighten_factor)
+        else
+            if bb:getInverse() == 1 then
+                -- MUL doesn't really work on a black background, so, switch to OVER if we're in software nightmode...
+                -- NOTE: If we do *not* invert the color here, it *will* get inverted by the blitter given that the target bb is inverted.
+                --       While not particularly pretty, this (roughly) matches with hardware nightmode, *and* how MuPDF renders highlights...
+                --       But it's *really* not pretty (https://github.com/koreader/koreader/pull/11044#issuecomment-1902886069), so we'll fix it ;p.
+                local c = Blitbuffer.ColorRGB32(color.r, color.g, color.b, 0xFF * self.highlight.lighten_factor):invert()
+                bb:blendRectRGB32(x, y, w, h, c)
+            else
+                bb:multiplyRectRGB(x, y, w, h, color)
+            end
+        end
     elseif drawer == "underscore" then
-        bb:paintRect(x, y + h - 1, w, Size.line.thick, Blitbuffer.COLOR_GRAY_4)
+        if not color then
+            color = Blitbuffer.COLOR_GRAY_4
+        end
+        if Blitbuffer.isColor8(color) then
+            bb:paintRect(x, y + h - 1, w, Size.line.thick, color)
+        else
+            bb:paintRectRGB32(x, y + h - 1, w, Size.line.thick, color)
+        end
     elseif drawer == "strikeout" then
+        if not color then
+            color = Blitbuffer.COLOR_BLACK
+        end
         local line_y = y + math.floor(h / 2) + 1
         if self.ui.paging then
             line_y = line_y + 2
         end
-        bb:paintRect(x, line_y, w, Size.line.medium, Blitbuffer.COLOR_BLACK)
+        if Blitbuffer.isColor8(color) then
+            bb:paintRect(x, line_y, w, Size.line.medium, color)
+        else
+            bb:paintRectRGB32(x, line_y, w, Size.line.medium, color)
+        end
     elseif drawer == "invert" then
         bb:invertRect(x, y, w, h)
     end
     if draw_note_mark then
+        if not color then
+            color = Blitbuffer.COLOR_BLACK
+        end
         if self.highlight.note_mark == "underline" then
-            bb:paintRect(x, y + h - 1, w, Size.line.medium, Blitbuffer.COLOR_BLACK)
+            if Blitbuffer.isColor8(color) then
+                bb:paintRect(x, y + h - 1, w, Size.line.medium, color)
+            else
+                bb:paintRectRGB32(x, y + h - 1, w, Size.line.medium, color)
```

Comment:
You draw the note marking underline with the same color as the highlighting itself, not visible.

Reply:
I imagine we probably actually want to keep those in black, always, not only the underline?

Reply:
Although... I can *sort* of see the sideline being in color cool?

Mildly less enthused about the icon, but I'm really, really not the target audience, so, no idea ^^

Reply:
Oh, the icon is already B&W only, so we're good ;p.